File: C++/range_sum_of_bst.cpp
// no. 938
#include "tree_node.h"
#include <vector>
using namespace std;

class Solution {
public:
  int rangeSumBST(TreeNode *root, int low, int high) {
    vector<TreeNode *> queue;
    queue.push_back(root);
    int ans = 0;
    while (!queue.empty()) {
      TreeNode *curr_node = queue.front();
      queue.erase(queue.begin());
      if (curr_node) {
        if (curr_node->val > high) {
          queue.push_back(curr_node->left);
        } else if (curr_node->val < low) {
          queue.push_back(curr_node->right);
        } else {
          ans += curr_node->val;
          queue.push_back(curr_node->left);
          queue.push_back(curr_node->right);
        }
      }
    }
    return ans;
  }
};

File: go/range_sum_of_bst.go
//no. 938
import tree

func rangeSumBST(root *TreeNode, low int, high int) int {
    var queue []*TreeNode
    queue = append(queue, root)
    ans := 0
    for{
        if len(queue) <= 0{
            break
        }
        var curr_node *TreeNode = queue[0]
        queue = queue[1:]
        if (curr_node != nil){
            if curr_node.Val > high{
                queue = append(queue, curr_node.Left)
            }else if curr_node.Val < low{
                queue = append(queue, curr_node.Right)
            }else{
                ans += curr_node.Val
                queue = append(queue, curr_node.Left)
                queue = append(queue, curr_node.Right)
            }
        }
    }
    return ans
}
